### PR TITLE
Helprank DB: baby steps

### PR DESCRIPTION
**Pull request by raymondng**
Ok finally started porting over existing HelpRank repos + (portion of) code to Heimdal!

Nothing is wired up at this point, so all the real actions still happen in shoebox. This pull requests consist of  some foundational pieces:

Repos (copy-and-paste; some renaming to avoid conflicts in dev mode)
HelpRank Commander/Controller (ported from Shoebox KeepInterner, etc.)
A couple end points (processKifiHit and processKeepAttribution)
(Lots of) Test setup and a couple of tests -- HelpRankCommanderTest being the main one (ported from KeepInternerTest)

Will deploy to make sure there are no guice issues, but other than that there are no real actions (yet).

Per discussion with @eishay the immediate goal is to make a quick migration, and then refactor/re-design things as needed in Heidmal -- we know there will be some refactoring/redesign needed to accomodate library/feed, take advantage of Heimdal's event tracking mechanism, better eliza endpoint, etc.

Next step will be to port more pieces (code + test) over and wire things up. And then copy the data over, and hope for the best ...

@yingjieMiao @eishay @stkem

### CODE REVIEW

File: kifi-backend/modules/common/conf/evolutions/shoebox/218.sql [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# HEIMDAL
```
Comment (stkem) [inline kifi-backend/modules/common/conf/evolutions/shoebox/218.sql:1]: Why two separate evolution files?

Comment (stkem) [inline kifi-backend/modules/common/conf/evolutions/shoebox/218.sql:1]: Sorry, three.

File: kifi-backend/modules/heimdal/conf/heimdal.routes [hunk 2]
```diff
@@ -32,4 +32,7 @@ GET  /internal/heimdal/user/delighted/time             @com.keepit.controllers.D
 POST /internal/heimdal/user/delighted/answer           @com.keepit.controllers.DelightedController.postDelightedAnswer()
 POST /internal/heimdal/user/delighted/cancel           @com.keepit.controllers.DelightedController.cancelDelightedSurvey()
 
+POST /internal/heimdal/helprank/processKifiHit         @com.keepit.controllers.HelpRankController.processKifiHit()
+POST /internal/heimdal/helprank/processKeepAttribution @com.keepit.controllers.HelpRankController.processKeepAttribution()
```
Comment (stkem) [inline kifi-backend/modules/heimdal/conf/heimdal.routes:36]: A bit confused by these endpoints. Heimdal already receives information about both keeps and clicks through the events. I thought that was kinda the whole point behind putting this on Heimdal.

File: kifi-backend/modules/common/conf/evolutions/shoebox/218.sql (hunk 1 above)
Comment (raymondng) [inline kifi-backend/modules/common/conf/evolutions/shoebox/218.sql:1]: Hmm ... more modular?

File: kifi-backend/modules/heimdal/app/com/keepit/commander/HelpRankCommander.scala [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    airbrake: AirbrakeNotifier,
+    clock: Clock,
+    kifiHitCache: KifiHitCache,
+    userKeepInfoRepo: UserBookmarkClicksRepo,
+    keepDiscoveryRepo: KeepDiscoveryRepo,
+    rekeepRepo: ReKeepRepo,
+    elizaClient: ElizaServiceClient,
+    shoeboxClient: ShoeboxServiceClient) extends Logging {
+
+  implicit val fj = ExecutionContext.fj
```
Comment (stkem) [inline kifi-backend/modules/heimdal/app/com/keepit/commander/HelpRankCommander.scala]: Please use the play default execution context. We should not be switching out such base components without some clear consensus strategy. 
It makes the code harder to reason about (we have file level execution contexts pretty much everywhere else, for instance) and performance issues harder to understand.

File: kifi-backend/modules/heimdal/conf/heimdal.routes (hunk 2 above)
Comment (raymondng) [inline kifi-backend/modules/heimdal/conf/heimdal.routes:36]: That comes later. See overall comment (people tend to ignore that ...)

FWIW -- I agree with you.

Comment (stkem) [inline kifi-backend/modules/heimdal/conf/heimdal.routes:36]: I did read the overall comment. If you are doing it anyway why spend the time to build it this way just to remove it again more or less immediately?

Comment (raymondng) [inline kifi-backend/modules/heimdal/conf/heimdal.routes:36]: Yes. I made the same argument.

The other argument is that we want things off shoebox now. We can always refactor/redesign later on. I am ok with that as well.

File: kifi-backend/modules/heimdal/app/com/keepit/commander/HelpRankCommander.scala [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  private def searchAttribution(userId: Id[User], keep: Keep): Seq[ReKeep] = {
+    implicit val dca = TransactionalCaching.Implicits.directCacheAccess
+    kifiHitCache.get(KifiHitKey(userId, keep.uriId)) map { hit =>
+      val res = db.readWrite { implicit rw =>
+        keepDiscoveryRepo.getDiscoveriesByUUID(hit.uuid) collect {
+          case c if rekeepRepo.getReKeep(c.keeperId, c.uriId, userId).isEmpty =>
+            val rekeep = ReKeep(keeperId = c.keeperId, keepId = c.keepId, uriId = c.uriId, srcUserId = userId, srcKeepId = keep.id.get, attributionFactor = c.numKeepers)
+            val saved = rekeepRepo.save(rekeep)
+            log.info(s"[searchAttribution($userId,${keep.uriId})] rekeep=$saved; click=$c")
+            saved
```
Comment (eishay) [inline kifi-backend/modules/heimdal/app/com/keepit/commander/HelpRankCommander.scala]: you could use `KestrelCombinator#tap` here as in 

``` scala
rekeepRepo.save(rekeep) tap { saved =>
  log.info(s"[searchAttribution($userId,${keep.uriId})] rekeep=$saved; click=$c")
}
```

File: kifi-backend/modules/heimdal/app/com/keepit/commander/HelpRankCommander.scala (hunk 3 above)
Comment (raymondng) [inline kifi-backend/modules/heimdal/app/com/keepit/commander/HelpRankCommander.scala]: As previously discussed, this allows for more visibility & monitoring, etc. Let's sync-up.

File: kifi-backend/modules/heimdal/app/com/keepit/commander/HelpRankCommander.scala (hunk 4 above)
Comment (raymondng) [inline kifi-backend/modules/heimdal/app/com/keepit/commander/HelpRankCommander.scala]: Sure.

**Comment by raymondng** [comment]
Per discussion with @stkem will proceed with merge. Your feedback is appreciated!

**Pull request merged**